core/src/environment.rs

```diff
@@ -19,43 +16,36 @@ use crate::metrics::{increment, sample};
 /// keys, and `V` are their value.
 ///
 /// The linked list is composed of the current layer and the previous layers.
-/// The current layer is stored as an `Rc<Hashmap>`. It is inserted in the previous layers
-/// by cloning.
-/// The insertion is made by trying to get the current layer as mutable using
-/// [`Rc::get_mut`]. If it can, it means it is the only owner of this layer
-/// in the environment, allowing it to mutate it. If it cannot, it means that
-/// the current has been cloned and inserted in the previous environment already,
-/// so it can safely be reset as a new hashmap.
-/// The previous layers are set in order from the most recent one to the oldest.
+/// The current layer is stored as an `Rc<Hashmap>`. It is inserted in the
+/// previous layers by cloning.
+///
+/// Insertions are made by copying-on-write: if the current layer has no other
```

Comment:
Unless I'm misunderstanding (I'll probably know at the end of the review), I don't think copy-on-write is a really good name, because the whole point of the environment is to never copy anything (beyond  `Rc`s maybe). It's more like "share-on-write"?